solarforecastarbiter/metrics/probabilistic.py

```diff
@@ -0,0 +1,210 @@
+"""Probablistic forecast error metrics."""
+
+import numpy as np
+
+__all__ = [
+    "brier_score",
+    "brier_skill_score",
+    "reliability",
+    "resolution",
+    "uncertainty",
+    "sharpness",
+]
+
+
+def brier_score(fx, fx_prob, obs):
+    """Brier Score (BS).
+
+    Parameters
+    ----------
+    fx : (n,) array_like
+        Forecasts (physical units).
+    fx_prob : (n,) array_like
+        Probability [%] associated with the forecasts.
+    obs : (n,) array_like
+        Observations (physical unit).
+
+    Returns
+    -------
+    score : float
+        The Brier Score.
+
+    """
+
+    # event: 0=did not happen, 1=did happen
+    o = np.zeros_like(obs)
+    o[obs <= fx] = 1.0
+
+    # forecast probabilities [-]
+    f = fx_prob / 100.0
+
+    BS = np.mean((f - o) ** 2)
+    return BS
+
+
+def brier_skill_score(fx, fx_prob, ref, ref_prob, obs):
+    """Brier Skill Score (BSS).
+
+        BSS = 1 - BS_f / BS_ref
+
+    where BS_fx is the evaluated forecast and BS_ref is a reference forecast.
+
+    Parameters
+    ----------
+    fx : (n,) array_like
+        Forecasts (physical units).
+    fx_prob : (n,) array_like
+        Probability [%] associated with the forecasts.
+    ref : (n,) array_like
+        Reference forecast (physical units).
+    ref_prob : (n,) array_like
+        Probability [%] associated with the reference forecast.
+    obs : (n,) array_like
+        Observations (physical unit).
+
+    Returns
+    -------
+    skill : float
+        The Brier Skill Score [-].
+
+    """
+    bs_fx = brier_score(fx, fx_prob, obs)
+    bs_ref = brier_score(ref, ref_prob, obs)
+    return 1.0 - bs_fx / bs_ref
+
+
+def reliability(fx, fx_prob, obs):
+    """Reliability (REL) of the forecast.
+
+    Parameters
+    ----------
+    fx : (n,) array_like
+        Forecasts (physical units).
+    fx_prob : (n,) array_like
+        Probability [%] associated with the forecasts.
+    obs : (n,) array_like
+        Observations (physical unit).
+
+    Returns
+    -------
+    REL : float
+        The reliability of the forecast, where a perfectly reliable forecast
+        has REL = 0.
+
+    """
+
+    # event: 0=did not happen, 1=did happen
+    o = np.zeros_like(obs)
+    o[obs <= fx] = 1.0
+
+    # forecast probabilities [-]
+    f = fx_prob / 100.0
+
+    # get unique forecast probabilities
+    if len(f) < 1000:
+        n_decimals = 1
+    else:
+        n_decimals = 2
+
+    f = np.around(f, decimals=n_decimals)
+    REL = 0.0
+    for f_i in np.unique(f):
+        N_i = len(f[f == f_i])  # no. forecasts per set
+        o_i = np.mean(o[f == f_i])    # mean event value per set
+        REL += N_i * (f_i - o_i) ** 2
```

Comment:
The implementation appears correct. This might be faster, and whether or not its more clear probably depends on the reader:

```python
rel = 0.0
for f_i, N_i in np.nditer(np.unique(f, return_counts=True)):
    o_i = np.mean(o[f == f_i])
    rel += N_i * (f_i - o_i) ** 2
rel /= len(f)
```

Reply:
I like your suggested revision and with the formula added to the docstring, I don't think the code is less clear than the other version. I'll revise accordingly.